teku/src/main/java/tech/pegasys/teku/cli/subcommand/RemoteSpecLoader.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.cli.subcommand;
+
+import java.net.URI;
+import java.time.Duration;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiConsumer;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import tech.pegasys.teku.infrastructure.exceptions.InvalidConfigurationException;
+import tech.pegasys.teku.infrastructure.logging.SubCommandLogger;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.SpecFactory;
+import tech.pegasys.teku.spec.config.SpecConfigLoader;
+import tech.pegasys.teku.sync.forward.singlepeer.RetryDelayFunction;
+import tech.pegasys.teku.validator.remote.apiclient.OkHttpClientAuthLoggingIntercepter;
+import tech.pegasys.teku.validator.remote.apiclient.OkHttpValidatorRestApiClient;
+
+class RemoteSpecLoader {
+  private static BiConsumer<Throwable, Integer> retryErrObserver =
+      (Throwable e, Integer attempt) -> {
+        var errMsg = String.format("Attempt %d failed: \n%s\n", attempt, e.getMessage());
+        SubCommandLogger.SUB_COMMAND_LOG.error(errMsg);
```

Comment:
We typically avoid using line endings in error messages as it can mess with logging setups.  I'd omit the attempt count as well.

Which does raise an interesting problem here because we're actually executing this before logging has been configured.  For the validator exit subcommand it doesn't have any complex options, but the validator-client subcommand uses our full logging support which may include a custom log4j config file.  So we'd really want this message to go through a logger, but can't because it's only later in the startup process that logging actually gets configured.

I think we will need to rework logging configuration so we can make it happen earlier in the process.  Since this isn't on by default yet let's leave that for a follow up PR.  Logging to system.err as this will do isn't unreasonable, just not ideal.

Reply:
Interesting, thanks for the context. I will change this to a simple system.err for now then

Reply:
Stick with `SubCommandLogger.error` for now - it is just delegating to `System.err` which is what I meant.  Most subcommands don't configure the logging system at all currently.